### DISCUSSION TITLE
How does one dynamically import optional dependencies?

### DISCUSSION THREAD

**Discussion by tgelu** [webpack/webpack#18027] [Q&A] [answered]
I cannot either figure it out from the docs or from other sources online the answer to this question:

**How can I dynamically import a package (eg. `import('abc')`) which is an optional peer dependency?**

I am building a library which has optional features that dynamically import the optional peer dependencies. The app consuming my library tries to build with webpack. But this will result in the build failing with `ModuleNotFoundError` if the optional peer dependency isn't there. But since the dependency is optional (defined in `package.json`) this is wrong. If I instruct webpack to ignore the import with a magic comment then the dependency isn't bundled in case it's there, so it will always be missing. I just can't find the middle ground where I can just:

```js
import('abc')
  .then(/* the peer dep is there */)
  .catch(/* do something in case the optional peer dep isn't there*/)
```

This seems like an obvious use-case to me, since package managers allow one to express the requirement of an *optional* peer dependency. Am I having a wrong expectation?

_I know that some libraries take the approach of exposing a different entry point for the optional feature but that's besides the point, because that doesn't relate to dynamic imports._

_I also know that this works if one wants to *require* the package in a try...catch, but that's also besides the point. (and not feasible in my case anyway)_

**Comment by alexander-akait** [comment]
Please read https://github.com/webpack/webpack/issues/11874, shortly - your browser versions (in browserslist or in the `target` optin in webpack.config.js) are not supported async function, but you ask webpack to generate them

**Reply by tgelu** [reply]
I looked at the linked issue but not sure I am able to follow the logic :)

I am not asking webpack to do anything it is not doing already, which is to dynamically import a known chunk at runtime. Except that it's OK if the module isn't there at build time. Allow me to have the build pass and the promise always fail at runtime. I don't see any other meaning in dynamically importing an optional module.

**Reply by alexander-akait** [reply]
Sorry, wrong question (mixed up the tabs) :smile:

**Reply by alexander-akait** [reply]
Just use https://webpack.js.org/configuration/externals/ for your library to keep the dynamic import and then in your application you can deside what you want:
- if you want to bundle it - just install it in the application
- if no or load from externals sources - setup `externals` in your application

Another solution - build multiple version and using the `"exports"` field to provide different versions (with and without optional peer dependency).

Also yes, you can use:
```js
async function loadMyModule(catchCallback, finallyCallback) {
  let loaded = false;

  try {
    loaded = await import('abc')
  } catch (err) {
    // Logic when it doesn't found
    await catchCallback();
  } finally {
    await finallyCallback();
  }
}
```

I want to say there are more different solution, it based no how you want to handle `optional peer dependency` logic

**Reply by tgelu** [reply]
Thank you for the reply!

This is very similar to what we do at the moment.

```ts
async function loadMyModule(catchCallback, finallyCallback) {
  let loaded = false;

  try {
    loaded = await import('abc')
  } catch (err) {
    // Logic when it doesn't found
    await catchCallback();
  } finally {
    await finallyCallback();
  }
}
```

But please help me understand if I am missing some webpack config or maybe something else, because this works *only* when `abc` is there. If it is not, then I get a build time error even with my import being in a try...catch:

```
ModuleNotFoundError: Module not found: Error: Can't resolve '@react-aria/focus'
```

Does this mean I must mark `abc` as external?

**Reply by alexander-akait** [reply]
Yes, you need to make `abc` external, because it is not a part of your build and can doesn't exist

**Reply by tgelu** [reply]
I see. And this leads me back to my initial point. If I make it an external then I won't be able to actually have webpack include it in my build in case it is there (since it's an optional dependency).

I will try to restate the initial question because I think I did not make my question clear enough and we're talking past each other :)

```js
try {
  var x = require('abc')
} catch (_) {
  // do something else now
}
```

1) If `abc` is present at build time, it will be included in the build and required at runtime. All good.
2) But the above require will **not** fail the build even if `abc` is not resolved. That will simply be handled in the catch at runtime. (I even think webpack replaces the require with a `throw`).

```js
try {
  var x = await import('abc')
} catch (_) {
  // do something else now
}
```

Now that I switched it to a dynamic `import` point 1) is still true:
1) If `abc` is present at build time, it will be included in the build and imported at runtime. All good.
2) But if `abc` is not there at build time, I can't even get the build to pass now, even if my import is in a try...catch. If I make it an external then point 1) won't be true anymore.

So point number 2) is different in the above two cases without any justification. I expected point 2) to be the same in both cases.

I hope this clarifies what I meant.

**Reply by alexander-akait** [reply]
I see the problem, how you setup `externals`? 

> But the above require will not fail the build even if abc is not resolved. That will simply be handled in the catch at runtime. (I even think webpack replaces the require with a throw).

Using:
```
externals: {
  abc: false
}
``` 

Generates:
```
WARNING in ./src/entry.js 3:10-24
Module not found: Error: Can't resolve 'abc' in '/home/akait/IdeaProjects/webpack/src'
```

for `require`, do you expected this for `import(...)`? i.e. a warning?

Note - webpack still generates output for `await import('abc')` case

**Reply by alexander-akait** [reply]
If yes - feel free to send a PR, the fix is easy, just put:
```js
dep.optional = !!parser.scope.inTry;
```

after https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportParserPlugin.js#L289

**Reply by sag1v** [reply]
@tgelu I agree with your point, i think both require() and import() should treat the same way when a module is missing.

I have the same use case (this is why i landed in this thread)

I am wondering, do you have any issues using the require() method and not dynamic import?

Also, did you notice any issues with tree-shake support of your library while using rquire() instead of static import?

Another observation, don't you have issues with node environment? This is not just a webpack thing, node also fails and needs the try catch approach (like when running jest for example)

**Reply by alexander-akait** [reply]
> @tgelu I agree with your point, i think both require() and import() should treat the same way when a module is missing.

Already fixed for any `import(...)` - https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportParserPlugin.js#L299 and https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportParserPlugin.js#L334

**Reply by sag1v** [reply]
> > @tgelu I agree with your point, i think both require() and import() should treat the same way when a module is missing.
> 
> Already fixed for any `import(...)` - https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportParserPlugin.js#L299 and https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportParserPlugin.js#L334

Thanks 🙏🏽@alexander-akait 
Since which version is this fixed?

**Reply by alexander-akait** [reply]
Yeah

**Reply by sag1v** [reply]
@alexander-akait I mean what version of webpack has this fix? 😃

**Reply by alexander-akait** [reply, marked as answer]
https://github.com/webpack/webpack/releases/tag/v5.90.2

> make dynamic import optional in try/catch

**Reply by tgelu** [reply]
thank you very much! 🙏

**Reply by sag1v** [reply]
It seems to work indeed, but it does produce a "side effect", it will create a chunk for each dynamic import.

Maybe we can disable the creating of the chunk with `/*.webpackMode: "eager" */` but this will bring back the errors and build failure.

**Reply by alexander-akait** [reply]
@sag1v Can you create reproducible test repo? `eager` mode should work

**Reply by sag1v** [reply]
I will try this weekend

**Reply by ericcornelissen** [reply]
I stumbled onto this following [a bug report for a similar issue in a library of mine](https://github.com/ericcornelissen/shescape/issues/2619) and while the solution discussed here largely works, I found one additional piece of configuration that might be necessary (see below). Also, I'm wondering why webpack doesn't just recognize that the dependency being imported is an optional dependency and automatically applies the recommended configuration?

The piece that's missing from this discussion is that, even if you mark it as external, webpack will emit a warning if the dependency is not installed. Which under some configurations is treated as a failure, thus requiring you to additionally configure webpack to accept the warning using [`ignoreWarnings`](https://webpack.js.org/configuration/other-options/#ignorewarnings).
